First draft of reduction script

SimonHeybrock opened the pull request:
See comments and TODOs in script.

SimonHeybrock committed:
First draft of reduction script

SimonHeybrock committed:
Fix to use LoadRKH since Load gives wrong unit/dim

SimonHeybrock committed:
Add wrappers for background and transmission handling

SimonHeybrock committed:
Fix issues with coord mismatches and add solid angle.

Works, provided that we change neutron.convert to *not* switch coords to
attributes.

SimonHeybrock committed:
More masking and background handling

OwenArnold committed:
use local path

Prevent conflicts every time we edit to work locally

OwenArnold committed:
Comparison against mantid in notebook

OwenArnold committed:
Pixel size as per bounding box output

SimonHeybrock committed:
Update sans/reduction.ipynb

SimonHeybrock committed:
Merge pull request #20 from scipp/minor_fixes

Make easier to work locally

SimonHeybrock committed:
Remove unused import

SimonHeybrock committed:
Fix transmission equation. Note duplicate arg to CalculateTransmission

SimonHeybrock committed:
Fix issues from rebinning "distribution"

SimonHeybrock committed:
Fix filename and variable name

SimonHeybrock committed:
Use out args for better performance

SimonHeybrock committed:
Remove wrong hack with units

Mantid's displayed units for distributions are misleading, what we had
was correct.

SimonHeybrock committed:
Removal of detector_ifo no longer required

SimonHeybrock committed:
Cleanup and support for reduction by wavelength band

SimonHeybrock commented on sans/reduction.ipynb:
```diff
@@ -0,0 +1,400 @@
+    "    return background\n",
+    "\n",
+    "def setup_transmission(data):\n",
+    "    incident_beam = extract_monitor_background(data['spectrum', 0:1], 40000.0*sc.units.us, 99000.0*sc.units.us)\n",
+    "    transmission = extract_monitor_background(data['spectrum', 3:4], 88000.0*sc.units.us, 98000.0*sc.units.us)\n",
```
@OwenArnold @nvaytet Note the hack here to use range-based slice `['spectrum', 0:1]`. What we actually want is `['spectrum', 0]`, but this drops the position coord and unit conversion will not work any more. What is the best way to address this, here and in general? Should we store a workspace containing only monitors as a dataset instead, so this changes from:
```python
data['spectrum', 0:1]
```
to
```python
data['monitor17']
```
?

SimonHeybrock replied:
Actually that does not work, because the positions coords of the monitors conflict when added to a `Dataset`.

SimonHeybrock replied:
Maybe it should simply be a dict of data arrays?

SimonHeybrock replied:
@OwenArnold @nvaytet The issue with the requirement to use range-slicing here is interesting, because it highlights a difference between xarray and scipp, and may point at a potential solution:

**xarray**
- Slicing without range in xarray *preserves* all coordinates (even those of dims that are sliced, they become scalar).
- Modifications via non-range slices works because xarray *compares only dimension-coords* in operations.

**scipp**
- Slicing without range *drops* coords along dims that are sliced.
- Modification via non-range slices works because (despite scipp comparing *all coords, including non-dimension-coords) there are no conflicting coords.

**Question**

What is the right approach? Obviously both work. The question is: Do we **need** comparison of non-dimension coords, i.e., is this a valuable safety net, or more of a hindrance (causing problems such as dropped position coord when extracting a single spectrum via slicing)?

OwenArnold replied:
From the sounds of it, the `scipp` approach seems like the correct one to me. What we would be doing otherwise is still having coords with some special case loss of integrity rules for non-dimension-coords?

I can't see the advantage of preserving the coord as a scalar when non-range slicing? Maybe this is useful for diagnostics down the line?

SimonHeybrock replied:
> I can't see the advantage of preserving the coord as a scalar when non-range slicing? Maybe this is useful for diagnostics down the line?

By far more than diagnostics. For example, to extract a single spectrum. Right now you lose the position information and will not be able to run unit conversion any more.

SimonHeybrock replied:
Have a look at http://xarray.pydata.org/en/stable/computation.html#coordinates. They have three examples there:

```python
# slicing x
arr[1] - arr[0] # no x coord
arr[0] + 1 # has x coord
arr[0] - arr[0] # has x coord,
```

In scipp, none of these have an x coord. The maybe more relevant case is the reverse though: Consider a 1D data array that represents a TOF monitor. It should have a `position` coord, such that unit conversion works. This is not a dimension coord, because `position` is not a dimension of the monitor. Scipp will however prevent binary operations in this case --- users are forced to change the coord to an attribute.

Now, I am wondering if non-dimension coords in xarray behave almost like attributes in scipp?

Writing this, I am wondering if this could solve some of our issues with unit conversions: The problem was that we have to switch, e.g., positions from coord to attributes when doing a conversion to wavelength, such that we can then normalize to a monitor (which would have a conflicting position):
```python
# data has dims=('spectrum', 'tof'), `position` is a non-dimension coord
data = sc.neutron.convert(data, 'tof', 'wavelength') # weird coord->attr juggling, partially broken
```

 Now, with xarrays behavior for non-dimension coords, our conversion could be:
```python
# data has dims=('position', 'tof'), `position` is a dimension coord
data = sc.neutron.convert(data, {'source-pos', 'sample-pos', 'position', 'tof'}, {'theta', 'wavelength'})
# data has dims=('theta', 'wavelength')
# new coord `theta` is dimension coord
# `position` is a non-dimension coord
```
Ignoring things like `source-position` and `sample-position` coords, this appears to solve some (not all) of the issues we had:
- No weird/surprising changes between coords and attributes.
- Normalizing to monitor works, since position is not a dimension coord.
- Theta (needed for `Q` later) is naturally there, no need to compute it based on position *attribute*.

Downsides/unclear:
- Adding, e.g., data and monitor to the same dataset will drop scalar position coord of monitor (if data is added first), `dataset['monitor']` will not show position.

OwenArnold committed:
q resolution steps

Basics implemenation based on math present in mantid TOFSANSResolutionByPixel. This is not yet producing the same results as Mantid. That will need to be invesigated next. Also the implementation presented here is very slow with no attention whatsoever applied to performance.

There are some additional units for scipp that need to be supported for this to work. namely inverse velociy units s/m.

OwenArnold committed:
Do not operate on coords fixes

OwenArnold committed:
Correct result against mantid

still unoptimized

OwenArnold committed:
Minor cleanup avoid some sq then sqrt ops

OwenArnold committed:
Introduce to latest revision

OwenArnold committed:
Merge branch 'loki-reduction' of github.com:scipp/ess into loki_q_resolution

OwenArnold committed:
Cleanup and use neutron.convert for mod calc

return dq sq
use neturon.convert to convert moderator data var from time spread to wavelength

OwenArnold committed:
rename l1 to l_collimation

OwenArnold committed:
pass in l_collimation

OwenArnold committed:
Remove comparision data loading

OwenArnold committed:
Remove hard-coded geometry values

SimonHeybrock committed:
Merge pull request #23 from scipp/loki_q_resolution

loki q resolution

SimonHeybrock committed:
Remove print
